src/server_manager/ui_components/outline-server-settings.html

```diff
@@ -311,6 +397,39 @@ <h3>[[localize('settings-metrics-header')]]</h3>
           input.invalid = false;
           this.$.keysPortSaveButton.disabled = false;
         }
+      },
+      _accessKeyDataLimitEnabledChanged: function(e) {
+        const wasDataLimitEnabled = this.isAccessKeyDataLimitEnabled;
+        const isDataLimitEnabled = e.detail.value;
+        if (isDataLimitEnabled === undefined || wasDataLimitEnabled === undefined) {
+          return;
+        } else if (isDataLimitEnabled === wasDataLimitEnabled) {
+          return;
+        }
+        if (isDataLimitEnabled) {
+          this._requestSetAccessKeyDataLimit();
+        } else {
+          this.fire('RemoveAccessKeyDataLimitRequested');
+        }
+      },
+      _handleAccessKeyDataLimitInputKeyDown: function(event) {
+        if (event.key === 'Escape') {
+          this.$.accessKeyDataLimitInput.value = this.accessKeyDataLimit.value;
+          this.$.accessKeyDataLimitInput.blur();
+        } else if (event.key === 'Enter') {
+          this.$.accessKeyDataLimitInput.blur();
+        }
+      },
+      _requestSetAccessKeyDataLimit: function() {
+        if (this.$.accessKeyDataLimitInput.invalid) {
+          return;
+        }
+        const value = Number(this.$.accessKeyDataLimitInput.value);
+        const unit = this.$.accessKeyDataLimitUnits.selected;
+        this.fire('SetAccessKeyDataLimitRequested', {limit: {value, unit}});
```

Comment:
If we pass `serverSettings: this` in these events then we don't have to call appRoot.getServerView in app.ts